### ISSUE TITLE
Feature Request: Lambda Provisioned Concurrency

### ISSUE THREAD

**Issue by bflad** [terraform-providers/terraform-provider-aws#11128]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Today, AWS has announced the release of Provisioned Concurrency for Lambda Functions. HashiCorp is proud to announce launch day support for this feature in Terraform. Provisioned Concurrency is a feature that keeps functions initialized and hyper-ready to respond in double-digit milliseconds. This is ideal for implementing interactive services, such as web and mobile backends, latency-sensitive microservices, or synchronous APIs.

When you invoke a Lambda function, the invocation is routed to an execution environment to process the request. When a function has not been used for some time, when you need to process more concurrent invocations, or when you update a function, new execution environments are created. The creation of an execution environment takes care of installing the function code and starting the runtime. Depending on the size of your deployment package, and the initialization time of the runtime and of your code, this can introduce latency for the invocations that are routed to a new execution environment. This latency is usually referred to as a “cold start”. For most applications this additional latency is not a problem. For some applications, however, this latency may not be acceptable.

When you enable Provisioned Concurrency for a function, the Lambda service will initialize the requested number of execution environments so they can be ready to respond to invocations.

Available in version 2.41.0 of the Terraform AWS Provider, a new `aws_lambda_provisioned_concurrency_config` resource can manage this new functionality. The resource documentation can be found [on the Terraform website](https://www.terraform.io/docs/providers/aws/r/lambda_provisioned_concurrency_config.html).

### New Resource(s)

- `aws_lambda_provisioned_concurrency_config`

### Terraform Configuration

```hcl
resource "aws_lambda_provisioned_concurrency_config" "alias_example" {
  function_name                     = aws_lambda_alias.example.function_name
  provisioned_concurrent_executions = 100
  qualifier                         = aws_lambda_alias.example.name
}

resource "aws_lambda_provisioned_concurrency_config" "function_example" {
  function_name                     = aws_lambda_function.test.function_name
  provisioned_concurrent_executions = 200
  qualifier                         = aws_lambda_function.test.version
}
```

### References

- Announcement Whats New Post: TBD
- [Announcement Blog](https://aws.amazon.com/blogs/aws/new-provisioned-concurrency-for-lambda-functions/)
- Lambda Documentation: TBD
- AWS Go SDK: TBD

**Comment by bflad** [comment]
A new `aws_lambda_provisioned_concurrency_config` resource has been merged and will release with version 2.41.0 of the Terraform AWS Provider, shortly.

**Comment by ewbankkit** [comment]
[Announcement](https://aws.amazon.com/about-aws/whats-new/2019/12/aws-lambda-announces-provisioned-concurrency/).
